internal/wasm/interpreter/interpreter.go

```diff
@@ -149,45 +150,54 @@ type interpreterOp struct {
 
 // Release implements wasm.Engine Release
 func (me *moduleEngine) Release() error {
-	for _, cf := range me.compiledFunctions {
+	// Release all the function instances declared in this module.
+	for _, cf := range me.compiledFunctions[me.importedFunctionCounts:] {
 		me.parentEngine.deleteCompiledFunction(cf.funcInstance)
 	}
 	return nil
 }
 
 // Compile implements internalwasm.Engine Compile
 func (e *engine) Compile(importedFunctions, moduleFunctions []*wasm.FunctionInstance) (me wasm.ModuleEngine, err error) {
-	compiledFunctions := make([]*compiledFunction, len(importedFunctions)+len(moduleFunctions))
-	for i, f := range importedFunctions {
+	modEngine := &moduleEngine{
+		compiledFunctions:      make([]*compiledFunction, 0, len(importedFunctions)+len(moduleFunctions)),
+		parentEngine:           e,
+		importedFunctionCounts: len(importedFunctions),
+	}
+
+	for _, f := range importedFunctions {
 		cf, ok := e.getCompiledFunction(f)
 		if !ok {
-			return nil, fmt.Errorf("uncompiled imported function: %s", f.Name)
+			return nil, fmt.Errorf("uncompiled imported function: %s.%s", f.Module.Name, f.Name)
 		}
-		compiledFunctions[i] = cf
+		modEngine.compiledFunctions = append(modEngine.compiledFunctions, cf)
 	}
 
-	ret := &moduleEngine{compiledFunctions: compiledFunctions, parentEngine: e}
-	for _, f := range moduleFunctions {
+	for i, f := range moduleFunctions {
 		var compiled *compiledFunction
 		if f.Kind == wasm.FunctionKindWasm {
 			ir, err := wazeroir.Compile(f)
 			if err != nil {
-				return nil, fmt.Errorf("failed to compile Wasm to wazeroir: %w", err)
+				_ = modEngine.Release()
+				return nil, fmt.Errorf("function[%d/%d] failed to lower to wazeroir: %w", i, len(moduleFunctions)-1, err)
```

Comment:
```suggestion
				// TODO(Adrian): extract Module.funcDesc so that errors here have more context
				return nil, fmt.Errorf("function[%d/%d] failed to lower to wazeroir: %w", i, len(moduleFunctions)-1, err)
```